### ISSUE TITLE
[New Feature] Offline Songs With Indexed DB

### ISSUE THREAD

**Issue by rakesh4real** [abhisheknaiidu/A-POP#3]
I can contribute for this feature. Kindly let me know if we can work on it.

**Comment by abhisheknaiidu** [comment]
@rakesh4real Yea sure, but how exactly are we adding this feature?

**Comment by rakesh4real** [comment]
I can write a module to write and read data from indexed DB. As I don't know much about design of your web app I may need your help integrating it

**Comment by abhisheknaiidu** [comment]
@rakesh4real Oh ok, sounds good to me, Yea I'll help, whenever there is a need of me 😄

**Comment by abhisheknaiidu** [comment]
@rakesh4real are you working? Let me know if I can do any help!

**Comment by TumoM** [comment]
@abhisheknaiidu I'd like to give this feature a shot. I'll make a pull request if I manage! 😇

**Comment by rakesh4real** [comment]
@abhisheknaiidu I apologize for not being able to work on the feature. There was a sudden change of plan at my work which needed my undiverted attention. 

@TumoM  Hi, if you are new to IndexedDB checkout 
[this gist](https://gist.github.com/rakesh4real/c04232c34a0602e39f45d8398370fc39). It will give you a jump start. Instead of saving strings, save string representation of binary data or do something similar to it

**Comment by TumoM** [comment]
> @abhisheknaiidu I apologize for not being able to work on the feature. There was a sudden change of plan at my work which needed my undiverted attention. 
> 
> @TumoM  Hi, if you are new to IndexedDB checkout 
> [this gist](https://gist.github.com/rakesh4real/c04232c34a0602e39f45d8398370fc39). It will give you a jump start. Instead of saving strings, save string representation of binary data or do something similar to it

@rakesh4real you're completely right, and we're on the same page with the binary idea, plain strings probably won't work. I did a little demo trying to pull only the audio from a YouTube video online, saving that to a Base64 string, and passing it into the Player Component (React Player) and it surprisingly worked. So now it's just time to sort out the logic within the app.

Good advice, thank you!

**Comment by abhisheknaiidu** [comment]
@rakesh4real It's fine!
and @TumoM thanks for showing interest, lemme me know if you want any help!

**Comment by TumoM** [comment]
> @rakesh4real It's fine!
> and @TumoM thanks for showing interest, lemme me know if you want any help!

Thanks @rakesh4real. I've just been experimenting with saving a song offline first without corrupting it, and passing that data to React Player.

@abhisheknaiidu gave a good gist for working with IndexDb. I'll get more done over the weekend.

I'll try implement the feature for the currently playing song, and get a working version going. I'll probably store an offline boolean in the store/state/context for each song , that runs a check against the IndexDb at startup and when a user presses play on a song. Otherwise it will fetch from the link. Tell me what you think about that flow @rakesh4real
